### DISCUSSION TITLE
Returning HTMX or JSON in one endpoint

### DISCUSSION THREAD

**Discussion by 3tilley** [litestar-org/litestar#2686] [Q&A] [answered]
### Summary

Hello,

I've just started using `litestar`, it seems cool!

I would like to serve JSON and HTMX responses from a similar endpoint, but without duplicating the code. I would also like to gate it on a header like `Accept: application/json`. Are there patterns for doing this? The only thing I can think of is:

```
def foo_logic(input):
    ...

@get("/foo")
def foo_get(request: Request):
    data = get_data_from_request(request)
    if data.header == "application/json":
        return foo_logic(data)
    else:
        return make_htmx(data)
```

This is a bit of a mess because I don't get free deserialisation on the input, and it's pretty confused in a type-sense as to what it's returning. I also have to have two different ways of deciphering that data, form-wise and data-wise.

I could have two separate functions, like `/api/foo` and `/foo` but that feels like a lot of boilerplate and work to keep them in sync.

How do other people deal with this?

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2685">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2685/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/2685/pledge.svg">
</picture>
</a>

**Comment by JacobCoffee** [comment, marked as answer]
> I could have two separate functions, like /api/foo and /foo but that feels like a lot of boilerplate and work to keep them in sync.

I agree, but this is how I deal with it personally at work because if I did these weird mixins I lose out of some features and clear concern separation. I hate it (the duplication) too, tbf.
